posthog/batch_exports/models.py

```diff
@@ -298,9 +307,40 @@ def workflow_id(self) -> str:
         return f"{self.batch_export.id}-Backfill-{start_at}-{end_at}"
 
     @property
-    def total_runs(self) -> int | None:
-        """Return the total number of runs for this backfill, based on the number of intervals."""
-        if not self.start_at or not self.end_at:
-            # just return None in this case since we don't have enough information to calculate the total runs
+    def total_expected_runs(self) -> int | None:
+        """Return the total number of expected runs for this backfill, based on the number of intervals."""
+        # if no start_at then it means we're backfilling all data in a single run
+        if self.start_at is None:
+            return 1
+
+        end_at = self.end_at
+        # if the backfill has no end_at then it means it's backfilling everything up to the 'present' (whatever that
+        # is defined as depends on whether the backfill is still running or not)
+        if not end_at and self.finished_at:
+            # if backfill has finished then we can use the finished_at as the approximate end_at
+            end_at = self.finished_at
+        elif not end_at and self.status == self.Status.RUNNING:
+            # if backfill is still running then we use the current time as the approximate end_at (it will obviously
+            # keep changing but is arguably better than returning nothing if a user wants to know a rough estimate of
+            # progress)
+            end_at = dt.datetime.now(tz=dt.UTC)
+        elif not end_at:
+            # we didn't always populated finished_at in the past, so probably don't have enough information
             return None
-        return (self.end_at - self.start_at) // self.batch_export.interval_time_delta
+        return ceil((end_at - self.start_at) / self.batch_export.interval_time_delta)
+
+    def get_finished_runs(self) -> int:
+        """Return the number of finished runs for this backfill.
+
+        This is primarily used to report progress of this backfill.
+        Note that this is just approximate since:
+        a) we don't know how many records are in each run of the backfill
+        b) some runs may have been cancelled or terminated (these are excluded since they may be retried manually and we
+            don't want to double count them)
+        """
+        return self.runs.filter(
+            status__in=[
+                BatchExportRun.Status.COMPLETED,
+                BatchExportRun.Status.FAILED,
+            ]
+        ).count()
```

Comment:
I don't like Django lets you do this. I was so confused looking for a `runs` attribute on this class. I personally don't think the loss in clarity is worth it to save the `BatchExportRun.objects.filter(backfill=self, ...)`

Reply:
But if it's a very common django pattern I'm OK accepting this.

Reply:
To be honest, I hadn't thought of that. But considering the fact that you found it confusing, I think it's worth making it more explicit like you suggest